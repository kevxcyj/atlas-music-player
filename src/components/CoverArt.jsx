import placeholder from '../assets/placeholder.svg';

const CoverArt = () => {
  return (
    <div className="w-full aspect-square bg-gray-200 flex items-center justify-center rounded-lg overflow-hidden dark:bg-gray-800">
      <img src={placeholder} alt="Cover Art" className="object-cover w-full h-full" />
    </div>
  );
};

export default CoverArt;