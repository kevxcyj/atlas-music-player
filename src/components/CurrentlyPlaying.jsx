import CoverArt from './CoverArt';
import SongTitle from './SongTitle';
import PlayControls from './PlayControls';
import VolumeControls from './VolumeControls';

const CurrentlyPlaying = () => {
  return (
    <div className="flex flex-col items-center justify-center p-6 space-y-6">
      <div className="w-64">
        <CoverArt />
      </div>
      <SongTitle />
      <PlayControls />
      <VolumeControls />
    </div>
  );
};

export default CurrentlyPlaying;