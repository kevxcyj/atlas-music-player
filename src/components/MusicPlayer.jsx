import CurrentlyPlaying from './CurrentlyPlaying';
import Playlist from './Playlist';

const MusicPlayer = () => {
  return (
    <div className="flex flex-col md:flex-row p-6 md:p-8 space-y-8 md:space-y-0 md:space-x-8 bg-bg-light dark:bg-bg-dark">
      <div className="flex-1">
        <CurrentlyPlaying />
      </div>
      <div className="flex-1">
        <Playlist />
      </div>
    </div>
  );
};

export default MusicPlayer;
