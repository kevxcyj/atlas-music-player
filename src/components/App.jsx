import React, { useState, useEffect } from 'react';
import MusicPlayer from './MusicPlayer';
import LoadingSkeleton from './LoadingSkeleton';
import '/src/index.css';

function App() {
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const timer = setTimeout(() => {
      setIsLoading(false);
    }, 2000);

    return () => clearTimeout(timer);
  }, []);

  return (
    <div className="flex flex-col min-h-screen bg-light dark:bg-bg-dark text-primary-text dark:text-bg-light">
      {isLoading ? <LoadingSkeleton /> : <MusicPlayer />}
    </div>
  );
}

export default App;
