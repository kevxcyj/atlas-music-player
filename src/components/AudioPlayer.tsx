import React, { useEffect, useRef } from 'react';
import { Song } from './MusicPlayer';

interface AudioPlayerProps {
  currentSong: Song | null;
  isPlaying: boolean;
  volume: number;
  playbackSpeed: number;
}

const AudioPlayer: React.FC<AudioPlayerProps> = ({ currentSong, isPlaying, volume, playbackSpeed }) => {
  const audioRef = useRef<HTMLAudioElement | null>(null);

  useEffect(() => {
    if (audioRef.current && currentSong) {
      audioRef.current.src = currentSong.url;
    }
  }, [currentSong]);

  useEffect(() => {
    if (audioRef.current) {
      if (isPlaying) {
        audioRef.current.play();
      } else {
        audioRef.current.pause();
      }
    }
  }, [isPlaying]);

  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.volume = volume;
    }
  }, [volume]);

  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.playbackRate = playbackSpeed;
    }
  }, [playbackSpeed]);

  return (
    <audio ref={audioRef} />
  );
};

export default AudioPlayer;