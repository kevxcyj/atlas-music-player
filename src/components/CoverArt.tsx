import placeholder from '../assets/placeholder.svg';
import { Song } from './MusicPlayer';

interface CoverArtProps {
  song: Song | null;
}

const CoverArt: React.FC<CoverArtProps> = ({ song }) => {
  const artworkSrc = song ? song.artwork : placeholder;

  return (
    <div className="w-full aspect-square bg-gray-200 flex items-center justify-center rounded-lg overflow-hidden dark:bg-gray-800">
      <img src={artworkSrc} alt="Cover Art" className="object-cover w-full h-full" />
    </div>
  );
};

export default CoverArt;