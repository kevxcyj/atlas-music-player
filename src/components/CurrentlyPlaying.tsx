import CoverArt from './CoverArt';
import SongTitle from './SongTitle';
import PlayControls from './PlayControls';
import VolumeControls from './VolumeControls';
import { Song } from './MusicPlayer';

interface CurrentlyPlayingProps {
  song: Song | null;
  playlist: Song[];
  isPlaying: boolean;
  setIsPlaying: (isPlaying: boolean) => void;
  playbackSpeed: number;
  setPlaybackSpeed: (speed: number) => void;
  isShuffling: boolean;
  setIsShuffling: (isShuffling: boolean) => void;
  volume: number;
  onVolumeChange: (newVolume: number) => void;
  playNextSong: () => void;
  playPreviousSong: () => void;
}

const CurrentlyPlaying: React.FC<CurrentlyPlayingProps> = ({ 
  song, 
  playlist,
  isPlaying, 
  setIsPlaying, 
  playbackSpeed, 
  setPlaybackSpeed, 
  isShuffling, 
  setIsShuffling, 
  volume, 
  onVolumeChange, 
  playNextSong, 
  playPreviousSong 
}) => {
  return (
    <div className="flex flex-col items-center justify-center p-6 space-y-6">
      <div className="w-full max-w-sm mx-auto">
        <CoverArt song={song} />
      </div>
      <SongTitle song={song} />
      <PlayControls 
        isPlaying={isPlaying}
        setIsPlaying={setIsPlaying}
        playbackSpeed={playbackSpeed}
        setPlaybackSpeed={setPlaybackSpeed}
        isShuffling={isShuffling}
        setIsShuffling={setIsShuffling}
        playNextSong={playNextSong}
        playPreviousSong={playPreviousSong}
        song={song}
        playlist={playlist}
      />
      <VolumeControls 
        volume={volume}
        onVolumeChange={onVolumeChange}
      />
    </div>
  );
};

export default CurrentlyPlaying;