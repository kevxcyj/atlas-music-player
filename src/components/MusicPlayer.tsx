import React, { useState, useEffect } from 'react';
import LoadingSkeleton from './LoadingSkeleton';
import CurrentlyPlaying from './CurrentlyPlaying';
import Playlist from './Playlist';
import AudioPlayer from './AudioPlayer';

export interface Song {
  id: number;
  title: string;
  artist: string;
  artwork: string;
  duration: number; // Changed to number
  url: string;
}

const MusicPlayer = () => {
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [playlist, setPlaylist] = useState<Song[]>([]);
  const [currentSong, setCurrentSong] = useState<Song | null>(null);
  const [isPlaying, setIsPlaying] = useState<boolean>(false);
  const [volume, setVolume] = useState<number>(50);
  const [playbackSpeed, setPlaybackSpeed] = useState<number>(1);
  const [isShuffling, setIsShuffling] = useState<boolean>(false);

  useEffect(() => {
    const fetchPlaylist = async () => {
      try {
        const response = await fetch('/api/v1/playlist');
        const data = await response.json();
        setPlaylist(data);
        if (data.length > 0) {
          setCurrentSong(data[0]);
        }
      } catch (error) {
        console.error("Failed to fetch playlist:", error);
      } finally {
        setIsLoading(false);
      }
    };

    fetchPlaylist();
  }, []);

  const handleSongSelect = (song: Song) => {
    setCurrentSong(song);
    setIsPlaying(true);
  };
  
  const handleVolumeChange = (newVolume: number) => {
      setVolume(newVolume);
  };

  const playPreviousSong = () => {
    if (!playlist || playlist.length === 0 || !currentSong) return;
    const currentIndex = playlist.findIndex(song => song.id === currentSong.id);
    if (currentIndex > 0) {
        setCurrentSong(playlist[currentIndex - 1]);
        setIsPlaying(true);
    }
  };

  const playNextSong = () => {
    if (!playlist || playlist.length === 0 || !currentSong) return;
    const currentIndex = playlist.findIndex(song => song.id === currentSong.id);
    if (isShuffling) {
        const randomIndex = Math.floor(Math.random() * playlist.length);
        setCurrentSong(playlist[randomIndex]);
    } else if (currentIndex < playlist.length - 1) {
        setCurrentSong(playlist[currentIndex + 1]);
    }
    setIsPlaying(true);
  };

  if (isLoading) {
    return <LoadingSkeleton />;
  }

  return (
    <div className="flex flex-col md:flex-row p-6 md:p-8 space-y-8 md:space-y-0 md:space-x-8 bg-bg-light dark:bg-bg-dark text-primary-text dark:text-bg-light">
      <div className="flex-1">
        <CurrentlyPlaying 
          song={currentSong}
          playlist={playlist}
          isPlaying={isPlaying} 
          setIsPlaying={setIsPlaying}
          playbackSpeed={playbackSpeed}
          setPlaybackSpeed={setPlaybackSpeed}
          isShuffling={isShuffling}
          setIsShuffling={setIsShuffling}
          volume={volume}
          onVolumeChange={handleVolumeChange}
          playPreviousSong={playPreviousSong}
          playNextSong={playNextSong}
        />
        <AudioPlayer 
          currentSong={currentSong}
          isPlaying={isPlaying}
          volume={volume / 100}
          playbackSpeed={playbackSpeed}
        />
      </div>
      <div className="flex-1">
        <Playlist 
          playlist={playlist} 
          currentSong={currentSong} 
          onSongSelect={handleSongSelect} 
        />
      </div>
    </div>
  );
};

export default MusicPlayer;