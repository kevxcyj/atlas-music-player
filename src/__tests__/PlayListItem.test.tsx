import { render } from '@testing-library/react';
import { expect, test } from 'vitest';
import PlaylistItem from '../components/PlayListItem';
import { Song } from '../components/MusicPlayer';

const mockSong: Song = {
  id: 1,
  title: 'Painted in Blue',
  artist: 'Soul Canvas',
  artwork: '',
  duration: 355,
  url: '',
};

test('PlaylistItem renders correctly when not selected', () => {
  const { container } = render(<PlaylistItem song={mockSong} isSelected={false} onSelect={() => {}} />);
  expect(container).toMatchSnapshot();
});

test('PlaylistItem renders correctly when selected', () => {
  const { container } = render(<PlaylistItem song={mockSong} isSelected={true} onSelect={() => {}} />);
  expect(container).toMatchSnapshot();
});