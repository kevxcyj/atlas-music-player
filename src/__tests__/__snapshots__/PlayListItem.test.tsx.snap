// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlaylistItem renders correctly when not selected 1`] = `
<div>
  <div
    class="flex items-center justify-between p-4 rounded-lg hover:bg-gray-100 dark:hover:bg-gray-800 cursor-pointer transition duration-300 ease-in-out transform hover:scale-105"
  >
    <div
      class="flex-1"
    >
      <h4
        class="text-primary-text dark:text-bg-light font-bold text-lg"
      >
        Painted in Blue
      </h4>
      <p
        class="text-gray-500 dark:text-gray-400 text-sm"
      >
        Soul Canvas
      </p>
    </div>
    <span
      class="text-primary-text dark:text-bg-light text-sm"
    >
      5:55
    </span>
  </div>
</div>
`;

exports[`PlaylistItem renders correctly when selected 1`] = `
<div>
  <div
    class="flex items-center justify-between p-4 rounded-lg bg-gray-200 dark:bg-gray-700"
  >
    <div
      class="flex-1"
    >
      <h4
        class="text-primary-text dark:text-bg-light font-bold text-lg"
      >
        Painted in Blue
      </h4>
      <p
        class="text-gray-500 dark:text-gray-400 text-sm"
      >
        Soul Canvas
      </p>
    </div>
    <span
      class="text-primary-text dark:text-bg-light text-sm"
    >
      5:55
    </span>
  </div>
</div>
`;
